### ISSUE TITLE
twisted docs about udp networking are not python 3 compatible

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8280]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)| @pawelmhm reported|
|-|-|
|Trac ID|trac#8280|
|Type|enhancement|
|Created|2016-04-09 07:35:28Z|
|Branch|https://github.com/twisted/twisted/tree/udpdocs3-8280|

twisted.internet.udp has been ported to Python 3 some time ago in [#6003](https://github.com/twisted/twisted/issues/6003), but seems like docs still use examples that are not compatible with Python 3. Worth fixing this

Attachments:

* [udp_docs.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/99b/99b0cd0c2c548f642d9e1e0a043803b609c3b4d8/5f3578e1b89c51b4a501f7efe173e0b037ab22a7.diff) (13624 bytes) - added by pawelmhm on 2016-04-09 11:36:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8280 8280
type__enhancement enhancement
reporter__pawelmhm pawelmhm
priority__normal normal
milestone__None None
branch__branches_udpdocs3_8280 branches/udpdocs3-8280
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__website website
keywords__None None
time__1460187328171344 1460187328171344
changetime__1460668562738338 1460668562738338
version__None None
owner__glyph glyph
cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_2"></a>|@pawelmhm commented|
|-|-|

I'm thinking about this page here https://twistedmatrix.com/documents/current/core/howto/udp.html

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_3"></a>|@pawelmhm removed owner|
|-|-|

Here's the patch. It updates code samples to work with Python 3 and 2. Corrects some notes and clarifies some points that seemed unclear. All code samples have been moved to /listings directory and imported from there so that anyone can run them easily. 

I was somewhat confused by multicast example. I "ported" it too and then tried testing. Example seems to work on first sight but if you look closer it is really unusual. First interesting thing is that commenting out "joinGroup()" call in MultiCastClient does not break anything. It should break all example IMO, because it means that socket is not added to multicast group. 

Other curiosity is that you dont need to start MultiCastServer to achieve effect that is described as effect of starting it. In other words you can start couple of clients alone and messages will be posted to each one of them. This is because each client is writing to same port address "228.0.0.5", 8005 and listening on same port: 8005. (BTW I switched all ports used to 9999 so that user dont have to remember all different ports in examples). 

Yet another anomaly is that "server: pong" message that should be sent only to sending client according to comment - is sent only to server itself. Comment in example says this message should be unicasted to "client" that sent "Ping", but this is not what I noticed.

I can try fixing this example if community considers it useful (but ATM I'm not sure how to fix it) or writing new example (this may be easier). If we think this example is not useful I would suggest removing it because it looks somewhat confusing as it stands. Of course please check my words here and make sure I understand this example correctly.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [47233]) Branching to udpdocs3-8280.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @glyph|
|-|-|

Thanks for your contribution, pawelmhm!

Replying to [pawelmhm](#note_3):
> Here's the patch. It updates code samples to work with Python 3 and 2. Corrects some notes and clarifies some points that seemed unclear. All code samples have been moved to /listings directory and imported from there so that anyone can run them easily.

Thanks for the thorough clean-up.

> I was somewhat confused by multicast example. I "ported" it too and then tried testing. Example seems to work on first sight but if you look closer it is really unusual. First interesting thing is that commenting out "joinGroup()" call in MultiCastClient does not break anything. It should break all example IMO, because it means that socket is not added to multicast group.

This is a good catch, and while it's unfortunate that the docs are still wrong, I think we should fix this as a separate issue.  So I intend to merge this as-is and then we can address these other issues.

> Other curiosity is that you dont need to start MultiCastServer to achieve effect that is described as effect of starting it. In other words you can start couple of clients alone and messages will be posted to each one of them. This is because each client is writing to same port address "228.0.0.5", 8005 and listening on same port: 8005. (BTW I switched all ports used to 9999 so that user dont have to remember all different ports in examples).

In the future, please try to make one change at a time.  In this case, I'm OK with this change, but one reason to go with different port numbers is that it makes it clear which examples are referring to each other and which aren't, by matching or not-matching port numbers.

> Yet another anomaly is that "server: pong" message that should be sent only to sending client according to comment - is sent only to server itself. Comment in example says this message should be unicasted to "client" that sent "Ping", but this is not what I noticed.

Sounds like another good fix.

> I can try fixing this example if community considers it useful (but ATM I'm not sure how to fix it) or writing new example (this may be easier). If we think this example is not useful I would suggest removing it because it looks somewhat confusing as it stands. Of course please check my words here and make sure I understand this example correctly. 

Please feel free to re-write the documentation completely, as long as you can clearly explain what was wrong with the old docs and how the new ones were better.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set status to `closed`|
|-|-|

(In [47236]) Merge udpdocs3-8280: port udp examples to python 3

Author: pawelmhm

Reviewer: glyph

Fixes: [#8280](https://github.com/twisted/twisted/issues/8280)

The documentation examples for UDP now work on Python 3.
